vault_azure_access_credentials crashes with subscription_id specified

Hi!

I have compiled the latest release of 3.4.0 locally and it is now actively validating the credentials, however when I attempt to specify a subscription_id and tenant_id, I get a crash.

### Terraform Version
Terraform v1.1.7
on linux_amd64
vault 3.4.0

### Affected Resource(s)
Please list the resources as a list, for example:
- data vault_azure_access_credentials

### Terraform Configuration Files
```hcl
data "vault_azure_access_credentials" "creds" {
  backend = "azure"
  role    = "edu-app"
  validate_creds = true
  num_sequential_successes = 5
  num_seconds_between_tests = 20
  max_cred_validation_seconds = 1200 // 20 minutes
  subscription_id = "xxxxxxx-xxxx-xxxx-xxxx-xxxxxxx" 
  tenant_id = "xxxxxxx-xxxx-xxxx-xxxx-xxxxxxx" 
}
```

│ Error: Plugin did not respond
│ 
│   with data.vault_azure_access_credentials.creds,
│   on main.tf line 25, in data "vault_azure_access_credentials" "creds":
│   25: data "vault_azure_access_credentials" "creds" {
│ 
│ The plugin encountered an error, and failed to respond to the plugin.(*GRPCProvider).ReadDataSource call. The plugin logs may contain more details.
╵

Stack trace from the terraform-provider-vault plugin:

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x30 pc=0xfd0884]

goroutine 25 [running]:
github.com/hashicorp/terraform-provider-vault/vault.azureAccessCredentialsDataSourceRead(0xc000254300, 0x1395d40, 0xc000626280, 0x1db2240, 0xc000100000)
        /mnt/c/POCs/terraform-provider-vault/vault/data_source_azure_access_credentials.go:200 +0x8c4
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).read(0xc0002970a0, 0x161cc68, 0xc0009f0800, 0xc000254300, 0x1395d40, 0xc000626280, 0x0, 0x0, 0x0)
        /root/go/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.10.0/helper/schema/resource.go:346 +0x1ee
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).ReadDataApply(0xc0002970a0, 0x161cc68, 0xc0009f0800, 0xc000254200, 0x1395d40, 0xc000626280, 0xc000626280, 0xc000254200, 0x0, 0x0)
        /root/go/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.10.0/helper/schema/resource.go:569 +0xfd
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ReadDataSource(0xc000291398, 0x161cc68, 0xc0009f0800, 0xc0009fe8c0, 0x13b2c93, 0x12, 0x0)
        /root/go/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.10.0/helper/schema/grpc_provider.go:1133 +0x518
github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server.(*server).ReadDataSource(0xc0001be600, 0x161cd10, 0xc0009f0800, 0xc000a208c0, 0x0, 0x0, 0x0)
        /root/go/pkg/mod/github.com/hashicorp/terraform-plugin-go@v0.5.0/tfprotov5/tf5server/server.go:478 +0x2f5
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ReadDataSource_Handler(0x134c9a0, 0xc0001be600, 0x161cd10, 0xc000a11770, 0xc000a0ea80, 0x0, 0x161cd10, 0xc000a11770, 0xc000a30780, 0x16c)
        /root/go/pkg/mod/github.com/hashicorp/terraform-plugin-go@v0.5.0/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:416 +0x214
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0001efdc0, 0x1628238, 0xc000582d80, 0xc000a32d80, 0xc000691650, 0x1d6a450, 0x0, 0x0, 0x0)
        /root/go/pkg/mod/google.golang.org/grpc@v1.41.0/server.go:1279 +0x52b
google.golang.org/grpc.(*Server).handleStream(0xc0001efdc0, 0x1628238, 0xc000582d80, 0xc000a32d80, 0x0)
        /root/go/pkg/mod/google.golang.org/grpc@v1.41.0/server.go:1608 +0xd0c
google.golang.org/grpc.(*Server).serveStreams.func1.2(0xc0002ce160, 0xc0001efdc0, 0x1628238, 0xc000582d80, 0xc000a32d80)
        /root/go/pkg/mod/google.golang.org/grpc@v1.41.0/server.go:923 +0xab
created by google.golang.org/grpc.(*Server).serveStreams.func1
        /root/go/pkg/mod/google.golang.org/grpc@v1.41.0/server.go:921 +0x1fd

Error: The terraform-provider-vault plugin crashed!

### Expected Behavior
The subscription_id and tenant_id should be able to be set.

### Actual Behavior
When I don't specify subscription_id and tenant_id it runs successfully.

### Steps to Reproduce
Please list the steps required to reproduce the issue, for example:
1. `terraform plan`

### Important Factoids
Tagging @benashz for awareness.

### References
Are there any other GitHub issues (open or closed) or Pull Requests that should be linked here? For example:
- [GH-1384](https://github.com/hashicorp/terraform-provider-vault/pull/1384)

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,3 +1,8 @@
+## 3.4.1 (Unreleased)
+BUGS:
+* `data/azure_access_credentials`: Fix panic when `tenant_id` and `subscription_id` are specified together; add new `environment` override field
+  ([#1391](https://github.com/terraform-providers/terraform-provider-vault/pull/1391)).
+
 ## 3.4.0 (March 24, 2022)
 FEATURES:
 * `data/azure_access_credentials` Add `subscription_id` and `tenant_id` fields to used during credential validation ([#1384](https://github.com/terraform-providers/terraform-provider-vault/pull/1384)).

--- a/vault/data_source_azure_access_credentials.go
+++ b/vault/data_source_azure_access_credentials.go
@@ -97,6 +97,13 @@ func azureAccessCredentialsDataSource() *schema.Resource {
 				Description: "The tenant ID to use during credential validation. " +
 					"Defaults to the tenant ID configured in the Vault backend",
 			},
+			"environment": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Description: `The Azure environment to use during credential validation.
+Defaults to the environment configured in the Vault backend.
+Some possible values: AzurePublicCloud, AzureGovernmentCloud`,
+			},
 		},
 	}
 }
@@ -197,13 +204,28 @@ func azureAccessCredentialsDataSourceRead(d *schema.ResourceData, meta interface
 	}
 
 	clientOptions := &arm.ClientOptions{}
-	if e, ok := config.Data["environment"]; ok && e.(string) != "" {
-		env, err := azure.EnvironmentFromName(e.(string))
+	var e string
+	if v, ok := d.GetOk("environment"); ok {
+		e = v.(string)
+	} else {
+		data, err := getConfigData()
+		if err != nil {
+			return err
+		}
+		if v, ok := data["environment"]; ok && v.(string) != "" {
+			e = v.(string)
+		}
+	}
+
+	if e != "" {
+		env, err := azure.EnvironmentFromName(e)
 		if err != nil {
 			return err
 		}
 
 		switch env.Name {
+		case "AzurePublicCloud":
+			clientOptions.Endpoint = arm.AzurePublicCloud
 		case "AzureChinaCloud":
 			clientOptions.Endpoint = arm.AzureChina
 		case "AzureGovernmentCloud":
@@ -233,7 +255,7 @@ func azureAccessCredentialsDataSourceRead(d *schema.ResourceData, meta interface
 			if pr.RawResponse.StatusCode == http.StatusUnauthorized {
 				return fmt.Errorf("validation failed, unauthorized credentials from Vault, err=%w", pager.Err())
 			}
-			log.Printf("[DEBUG] Provider Client List response %+v", pr)
+			log.Printf("[DEBUG] Provider Client List response %+v", pr.RawResponse)
 		}
 
 		if pager.Err() == nil {
@@ -247,7 +269,7 @@ func azureAccessCredentialsDataSourceRead(d *schema.ResourceData, meta interface
 				return fmt.Errorf("validation failed, giving up err=%w", pager.Err())
 			}
 
-			log.Printf("[ERROR] Credential validation failed with %v, retrying in %s", pager.Err(), delay)
+			log.Printf("[WARN] Credential validation failed with %v, retrying in %s", pager.Err(), delay)
 			successCount = 0
 		}
 		time.Sleep(delay)

--- a/website/docs/d/azure_access_credentials.html.md
+++ b/website/docs/d/azure_access_credentials.html.md
@@ -87,6 +87,10 @@ to 300.
 * `tenant_id` - (Optional) The tenant ID to use during credential validation.
    Defaults to the tenant ID configured in the Vault `backend`.
 
+* `environment` - (Optional) The Azure environment to use during credential validation.
+  Defaults to the environment configured in the Vault backend.
+  Some possible values: `AzurePublicCloud`, `AzureGovernmentCloud`
+
 ## Attributes Reference
 
 In addition to the arguments above, the following attributes are exported:

